File: tests/machine/x/typescript/group_by_sort.ts
const items = [
  {cat: "a", val: 3},
  {cat: "a", val: 1},
  {cat: "b", val: 5},
  {cat: "b", val: 2}
];
const grouped = (() => {
  const _tmp1: Array<{ cat: any; total: number }> = [];
  const groups = {};
  for (const i of items) {
    const _k = JSON.stringify(i.cat);
    let g = groups[_k];
    if (!g) { g = []; g.key = i.cat; g.items = g; groups[_k] = g; }
    g.push(i);
  }
  for (const _k in groups) {
    const g = groups[_k];
    _tmp1.push({item: {
  cat: g.key,
  total: (g.map((x) => x.val).reduce((a,b)=>a+b,0))
}, key: (-(g.map((x) => x.val).reduce((a,b)=>a+b,0)))});
  }
  _tmp1 = _tmp1.sort((a,b)=> a.key < b.key ? -1 : a.key > b.key ? 1 : 0).map(x=>x.item);
  return _tmp1;
})()
;
console.log(grouped);

File: tests/machine/x/swift/group_by_sort.swift
struct CatVal: Equatable {
    var cat: String
    var val: Int
}

var items = [CatVal(cat: "a", val: 3), CatVal(cat: "a", val: 1), CatVal(cat: "b", val: 5), CatVal(cat: "b", val: 2)]
var grouped = { () -> [Any] in
    var _groups: [String:[CatVal]] = [:]
    for i in items {
        let _k = i.cat
        _groups[_k, default: []].append(i)
    }
    var _tmp: [(key: String, items: [CatVal])] = []
    for (k, v) in _groups {
        _tmp.append((key: k, items: v))
    }
    _tmp.sort { String(describing: $0.items.map { x in x.val }.reduce(0, +)) > String(describing: $1.items.map { x in x.val }.reduce(0, +)) }
    return _tmp.map { g in ["cat": g.key, "total": g.items.map { x in x.val }.reduce(0, +)] }
}()
print(grouped)
